### PR TITLE
create DocumentManager

### PR DESCRIPTION
**Pull request by lead8964878**
create DocumentManager

### CODE REVIEW

File: src/DocumentManager.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+/** @brief Document Manager */
+class DocumentManager {
+private:
+    /** @brief Map recording id and DocumentSession. */
+    std::map<id, DocumentSession*> docMap_;
+
+public:
+    DocumentManager();
+
```
Comment (YuanWeiChen16) [inline src/DocumentManager.h:42]: maybe need add destructor

File: src/DocumentManager.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    int id;
+    /** @brief Modify history list. */
+    std::vector<Revision> history;
+} Document;
+
+/** @brief Document Manager */
+class DocumentManager {
+private:
+    /** @brief Map recording id and DocumentSession. */
+    std::map<id, DocumentSession*> docMap_;
```
Comment (YuanWeiChen16) [inline src/DocumentManager.h:38]: all lowercase,like doc_map_ or doc_map
reference https://google.github.io/styleguide/cppguide.html#Variable_Names

File: src/DocumentManager.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+/** @brief Store document modify revision */
+typedef struct{
+    /** @brief Who made this revision. */
+    std::string userId;
+    /** @brief Relative insert position from document start and end. */
+    int startPosition;
+    int endPosition;
+    /** @brief insertData and deleteLength will only exist one at the same time. */
+    std::string insertData;
+    int deleteLength;
```
Comment (kungyao) [inline src/DocumentManager.h:24]: Please naming with all lower case.
https://google.github.io/styleguide/cppguide.html#Variable_Names

File: src/DocumentManager.h (hunk 2 above)
Comment (john0312) [reply src/DocumentManager.h:38]: Yes. This is a reasonable review.

Comment (john0312) [reply src/DocumentManager.h:38]: Also note, `id` doesn't seemed to be a type. Does this build?

File: src/DocumentManager.h (hunk 1 above)
Comment (john0312) [reply src/DocumentManager.h:42]: Destructor is not always needed. It is recommended to note why it is needed, i.e. something hints that cleanup is needed.

File: src/DocumentManager.h (hunk 3 above)
Comment (john0312) [reply src/DocumentManager.h:24]: Yes, this is a reasonable review.

File: src/DocumentManager.h (hunk 2 above)
Comment (lead8964878) [reply src/DocumentManager.h]: ok, `id` misstyped

File: src/DocumentManager.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+class DocumentSession;
+
+/** @brief Store document modify revision */
+typedef struct{
+    /** @brief Who made this revision. */
+    std::string user_id;
+    /** @brief Relative insert position from document start and end. */
+    int start_position;
+    int end_position;
+    /** @brief insertData and deleteLength will only exist one at the same time. */
```
Comment (john0312) [inline src/DocumentManager.h:22]: How do you tell if we look at `insert_data` or `delete_length`?

File: src/DocumentManager.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    std::string insert_data;
+    int delete_length;
+} Revision;
+
+typedef struct{
+    /** @brief Document id. */
+    int id;
+    /** @brief Modify history list. */
+    std::vector<Revision> history;
```
Comment (john0312) [inline src/DocumentManager.h:31]: Maybe add constructor to hint reader that this needs initialization. (std::vector requires initialization)

**Review by kungyao — APPROVED** [review]
Viewed

**Review by YuanWeiChen16 — APPROVED** [review]

File: src/DocumentManager.h (hunk 4 above)
Comment (lead8964878) [reply src/DocumentManager.h:28]: added is_insert to distinguish insert or delete revision

**Pull request merged**